engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java

```diff
@@ -0,0 +1,274 @@
+   *   <li>Zoned date time, such as '2011-12-03T10:15:30+01:00[Europe/Paris]' with optional region
+   *       id in square brackets.
+   * </ul>
+   *
+   * @param text the string to parse.
+   * @return parsed ZonedDateTime instance wrapped in EnsoDateTime.
+   */
+  @Builtin.Method(
+      name = "parse_builtin",
+      description = "Constructs a new DateTime from text with optional pattern")
+  @Builtin.Specialize
+  @Builtin.WrapException(
+      from = DateTimeParseException.class,
+      to = PolyglotError.class,
+      propagate = true)
+  public static EnsoDateTime parse(String text) {
+    TemporalAccessor time = TIME_FORMAT.parseBest(text, ZonedDateTime::from, LocalDateTime::from);
+    if (time instanceof ZonedDateTime) {
+      return new EnsoDateTime((ZonedDateTime) time);
+    } else if (time instanceof LocalDateTime) {
+      return new EnsoDateTime(((LocalDateTime) time).atZone(ZoneId.systemDefault()));
+    }
+    throw new DateTimeException("Text '" + text + "' could not be parsed as Time.");
+  }
+
+  @Builtin.Method(
+      name = "new_builtin",
+      description = "Constructs a new Date from a year, month, and day")
+  @Builtin.WrapException(from = DateTimeException.class, to = PolyglotError.class, propagate = true)
+  public static EnsoDateTime create(
+      long year,
+      long month,
+      long day,
+      long hour,
+      long minute,
+      long second,
+      long nanosecond,
+      EnsoZone zone) {
+    return new EnsoDateTime(
+        ZonedDateTime.of(
+            Math.toIntExact(year),
+            Math.toIntExact(month),
+            Math.toIntExact(day),
+            Math.toIntExact(hour),
+            Math.toIntExact(minute),
+            Math.toIntExact(second),
+            Math.toIntExact(nanosecond),
+            zone.asTimeZone()));
+  }
+
+  @Builtin.Method(description = "Gets the year")
+  public long year() {
+    return dateTime.getYear();
+  }
+
+  @Builtin.Method(description = "Gets the month")
+  public long month() {
+    return dateTime.getMonthValue();
+  }
+
+  @Builtin.Method(description = "Gets the day")
+  public long day() {
+    return dateTime.getDayOfMonth();
+  }
+
+  @Builtin.Method(description = "Gets the hour")
+  public long hour() {
+    return dateTime.getHour();
+  }
+
+  @Builtin.Method(description = "Gets the minute")
+  public long minute() {
+    return dateTime.getMinute();
+  }
+
+  @Builtin.Method(description = "Gets the second")
+  public long second() {
+    return dateTime.getSecond();
+  }
+
+  @Builtin.Method(description = "Gets the nanosecond")
+  public long nanosecond() {
+    return dateTime.getNano();
+  }
+
+  @Builtin.Method(name = "zone", description = "Gets the zone")
+  public EnsoZone zone() {
+    return new EnsoZone(dateTime.getZone());
+  }
+
+  @Builtin.Method(description = "Return the number of seconds from the Unix epoch.")
+  public long toEpochSeconds() {
```

Comment:
My [recent experience](https://github.com/enso-org/enso/pull/3638/files#diff-07942731a467d64d7df16571e777af374f8b72159502eae8a436ba7addca768d) with `@TruffleBoundary` suggests that many of these methods shall be behind the boundary.

Reply:
I can add some, but at the moment without any gatekeepers it feels like a slightly random process.